### PR TITLE
sstables: Fix incremental_reader_selector (for range reads) with tablets

### PR DESCRIPTION
**Pull request by raphaelsc**
sstables: Fix incremental_reader_selector (for range reads) with tablets

incremental_reader_selector is the mechanism for incremental comsumption
of disjoint sstables on range reads.

tablet_sstable_set was implemented, such that selector is efficient with
tablets.

The problem is selector is vnode addicted and will only consider a given
set exhausted when maximum token is reached.

With tablets, that means a range read on first tablet of a given shard
will also consume other tablets living in the same shard. That results
in combined reader having to work with empty sstable readers of tablets
that don't intersect with the range of the read. It won't cause extra
I/O because the underlying sstables don't intersect with the range of
the read. It's only unnecessary CPU work, as it involves creating
readers (= allocation), feeding them into combined reader, which will
in turn invoke the sstable readers only to realize they don't have any
data for that range.

With 100k tablets (ranges), and 100 tablets per shard, and ~5 sstables
per tablet, there will be this amount of readers (empty or not):
```
    (100k * ((100^2 + 100) / 2) * avg_sstable_per_tablet=5) = ~2.5 billions.
```

~5000 times more readers, it can be quite significant additional cpu
work, even though I/O dominates the most in scans. It's an inefficiency
that we rather get rid of.

The behavior can be observed from logs (there's 1 sstable for each of
4 tablets, but note how readers are created for every single one of
them when reading only 1 tablet range):
```
table - make_reader_v2 - range=(-inf, {-4611686018427387905, end}]
    incremental_reader_selector - create_new_readers(null): selecting on pos {minimum token, w=-1}
    sstable - make_reader - reader on (-inf, {-4611686018427387905, end}] for sst 3gfx_..._34qn42... that has range [{-9151620220812943033, start},{-4813568684827439727, end}]
    incremental_reader_selector - create_new_readers(null): selecting on pos {-4611686018427387904, w=-1}
    sstable - make_reader - reader on (-inf, {-4611686018427387905, end}] for sst 3gfx_..._368nk2... that has range [{-4599560452460784857, start},{-78043747517466964, end}]
    incremental_reader_selector - create_new_readers(null): selecting on pos {0, w=-1}
    sstable - make_reader - reader on (-inf, {-4611686018427387905, end}] for sst 3gfx_..._38lj42... that has range [{851021166589397842, start},{3516631334339266977, end}]
    incremental_reader_selector - create_new_readers(null): selecting on pos {4611686018427387904, w=-1}
    sstable - make_reader - reader on (-inf, {-4611686018427387905, end}] for sst 3gfx_..._3dba82... that has range [{5065088566032249228, start},{9215673076482556375, end}]
```

Fix is about making sure the tablet set won't select past the
supplied range of the read.

### CODE REVIEW

File: sstables/sstable_set.cc [hunk 1]
```diff
@@ -852,6 +857,14 @@ class incremental_reader_selector : public reader_selector {
 
         return {};
     }
+
+    // Can be false-positive but never false-negative!
+    virtual bool has_new_readers(const std::optional<dht::ring_position_view>& pos) const noexcept override {
+#ifndef SCYLLA_BUILD_MODE_RELEASE
+        assert(!pos || dht::ring_position_tri_compare(*_s, *pos, dht::ring_position_view::for_range_end(*_pr)) <= 0);
+#endif
```
Comment (denesb) [inline sstables/sstable_set.cc:865]: Use `on_internal_error()` instead, and remove the `#ifndef`.

File: sstables/sstable_set.cc [hunk 2]
```diff
@@ -852,6 +857,14 @@ class incremental_reader_selector : public reader_selector {
...
+        return !end_of_stream() && (!pos || dht::ring_position_tri_compare(*_s, *pos, _selector_position) >= 0);
```
Comment (denesb) [inline sstables/sstable_set.cc:866]: Looks like the logic is the same here as in the base class, the only difference is what pos we consider as `max`. S maybe we just need a `virtual bool is_max_pos()` and keep `has_new_readers()` non-virtual.

**Comment by raphaelsc** [comment]
v2:
- instead exclude tablets in tablet set that don't intersect with range of read. previous approach breaks interaction of compound set and incremental selector. It's safer to restrict change to tablet set, as disjointness can be assumed across tablets.
- updated commit message.

**Comment by raphaelsc** [comment]
@denesb please re-review.

**Comment by raphaelsc** [comment]
to fix clang-tidy issue, will have to rebase once https://github.com/scylladb/scylladb/pull/18596 lands on master

**Review by denesb — APPROVED** [review]
LGTM, but @bhalevy please also review because I'm not that familiar with sstable set internals anymore.

**Comment by raphaelsc** [comment]
v3:
rebased

File: sstables/sstable_set.cc [hunk 3]
```diff
@@ -206,14 +206,19 @@ sstable_set::incremental_selector::~incremental_selector() = default;
 sstable_set::incremental_selector::incremental_selector(sstable_set::incremental_selector&&) noexcept = default;
 
 sstable_set::incremental_selector::selection
-sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
-    if (!_current_range_view || !_current_range_view->contains(pos, _cmp)) {
-        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(pos);
+sstable_set::incremental_selector::select(selector_pos s) const {
+    if (!_current_range_view || !_current_range_view->contains(s.pos, _cmp)) {
+        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(s);
```
Comment (bhalevy) [inline sstables/sstable_set.cc:211]: ring_position_view is 24 bytes in size and the additional pointer to the range is 8 bytes long, so well be copying 32 bytes.  That seems a bit much.  Can we pass a reference to the selector_pos instead?
Those functions are synchronous, so it shouldn't fall out scope.

File: sstables/sstable_set.cc [hunk 4]
```diff
@@ -206,14 +206,19 @@ sstable_set::incremental_selector::~incremental_selector() = default;
...
         _current_range_view = _current_range->transform([] (const dht::ring_position& rp) { return dht::ring_position_view(rp); });
     }
     return {_current_sstables, _current_next_position};
 }
 
+sstable_set::incremental_selector::selection
+sstable_set::incremental_selector::select(const dht::ring_position_view& s) const {
+    return select(selector_pos{s});
```
[the remark this answers is not in the document]
Comment (raphaelsc) [reply sstables/sstable_set.cc:219]: alright.

File: sstables/sstable_set.cc (hunk 3 above)
Comment (raphaelsc) [reply sstables/sstable_set.cc:211]: Ok, you're asking for incremental_selector_impl::select() to receive 'const selector_pos&' instead? Yes, I can do that.

Comment (bhalevy) [reply sstables/sstable_set.cc:211]: yes, thanks

Comment (raphaelsc) [reply sstables/sstable_set.cc:211]: thank you.

**Comment by raphaelsc** [comment]
v4:
- incremental_selector::select(const dht::ring_position_view&) is moved to header
- pass selector_pos to incremental_selector_impl::select() as const ref.

**Review by bhalevy — APPROVED** [review]
LGTM

File: sstables/sstable_set.cc [hunk 5]
```diff
@@ -820,7 +820,7 @@ class incremental_reader_selector : public reader_selector {
         auto readers = std::vector<flat_mutation_reader_v2>();
 
         do {
-            auto selection = _selector->select(_selector_position);
+            auto selection = _selector->select({_selector_position, _pr});
```
Comment (avikivity) [inline sstables/sstable_set.cc:823]: Ah, we already sent _pr via create_reader(), but the next tablet's sstables would just reject it.

**Pull request closed without merging**